Thread management

akinsey opened the pull request:
This should add fully functional thread lock and sticky with testing

akinsey committed:
feat(thread-lock): add db method to lock thread

akinsey committed:
refactor(thread-lock): wip thread lock route, without authorizations

akinsey committed:
feat(thread-lock): implement authorizations for thread lock

akinsey committed:
test(data-case): update data case to have different authenticated roles

akinsey committed:
test(thread-lock): wip implement thread lock tests

akinsey committed:
test(thread-lock): wip finish thread lock tests

akinsey committed:
fix(authorizations): preload user roles when querying thread first post

akinsey committed:
Merge branch 'config-refactor' into thread-management

akinsey committed:
Merge remote-tracking branch 'origin/config-refactor' into thread-management

akinsey committed:
test(thread-lock): complete thread lock tests

unenglishable reviewed (commented):
pending merge on config-refactor

akinsey committed:
feat(sticky): implement sticky thread db function

akinsey committed:
feat(sticky-json): add json file for new sticky route

akinsey committed:
feat(sticky): implement sticky thread route with auth

akinsey committed:
test(thread): add doctest for thread

akinsey committed:
test(thread-sticky): implement testing for thread sticky

akinsey committed:
style(format): run mix format

akinsey committed:
refactor(thread-purge): wip implement thread purge db route

akinsey committed:
feat(profile): decrement post count db function for user profile

akinsey committed:
feat(purge-thread): implement db function for purging a thread

akinsey committed:
Merge remote-tracking branch 'origin/config-refactor' into purge

akinsey committed:
fix(purge): resolve issues with purge db function

akinsey committed:
fix(thread-sticky): create new private acl check for sticky thread

akinsey committed:
Merge remote-tracking branch 'origin/config-refactor' into purge

akinsey committed:
refactor(purge): wip implement purge

akinsey committed:
fix(purge): credo error, use Enum each instead of Enum map, run format

akinsey committed:
feat(purge): implement email template for purging thread

akinsey committed:
feat(users): implement query function to get email data via list of user ids

akinsey committed:
feat(purge-email): email users of thread when purging thread

akinsey committed:
fix(purge-email): send correct wording to thread author not person authenticated

akinsey committed:
refactor(purge): return board name with purge query so it doesnt have to be queried by moderation log

akinsey committed:
refactor(mod-log): refactor variable name in mod log template for thread purge

akinsey committed:
fix(ip): remove ipv6 prefix on ipv4 addresses

akinsey committed:
feat(mod-log-purge): add record to mod log when purging thread

akinsey committed:
test(mod-log): fix modlog test, update params for purge json test

akinsey committed:
test(thread-purge): implement test for thread purge

akinsey committed:
test(purge): check post count after purge

akinsey committed:
fix(dialyzer): fix type for threads to add virtual fields that purge function returns

akinsey committed:
feat(watch-thread): impement initial watch thread api route

akinsey committed:
fix(watch-thread): update create method to have consistent return value

akinsey committed:
fix(decode-plug): sane return value when method does not match @methods list

akinsey committed:
fix(watch-thread): update watchThread delete function to take user map

akinsey committed:
feat(watch-thread): implement unwatch thread route

akinsey committed:
test(thread-controller): add global mod user return to setup method for use in tests

akinsey committed:
test(thread-contoller): unlock thread test

akinsey committed:
test(thread-controller): fix thread test language

akinsey committed:
test(thread-contoller): unsticky thread test

akinsey committed:
test(thread-controller): implement test for watch thread

akinsey committed:
test(thread-controller): implement test for thread unwatch.

akinsey committed:
test(thread-controller): update test using format

unenglishable committed:
Merge pull request #101 from epochtalk/watch-thread

Watch thread

unenglishable committed:
Merge remote-tracking branch 'origin/main' into purge

unenglishable committed:
Merge pull request #100 from epochtalk/purge

Purge

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :global_mod
+    test "given thread that authenticated user moderates, does lock poll", %{
```
does lock thread

unenglishable replied:
suggestion "does lock poll" -> "locks thread"

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :global_mod
+    test "given thread that authenticated user moderates, does unlock poll", %{
```
does unlock thread

unenglishable replied:
suggestion "does unlock poll" -> "unlocks thread"

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -21,14 +30,25 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+
     {
       :ok,
+      user: user,
```
don't need to return user here, not modified

akinsey replied:
This is used on line 557 I believe

"given thread that authenticated user moderates, does purge thread"

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -21,14 +30,25 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
     admin_threads = build_list(3, :thread, board: admin_board, user: admin_user)
     super_admin_threads = build_list(3, :thread, board: super_admin_board, user: super_admin_user)
 
+    thread = build(:thread, board: board, user: user)
```
why not reuse threads, admin_threads, super_admin_threads?

unenglishable replied:
the one different use case is `admin_created_thread`, since it uniquely has a thread created by admin user under non-admin board

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+  describe "lock/2" do
+    test "when unauthenticated, returns Unauthorized error", %{
+      conn: conn,
+      thread: %{post: %{thread_id: thread_id}}
```
reusing threads eg:

threads: [%{post: %{thread_id: thread_id}} | _]

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :banned
+    test "given banned authenticated user, throws InvalidPermission forbidden error", %{
```
"given" should be for data, please use "when logged in as..." or "when authenticated with..." to specify authenticated user

this applies to all other tests in the file as well, please check them

unenglishable replied:
eg:  "when authenticated with banned user, throws InvalidPermission forbidden error"

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :mod
+    test "given admin thread and insufficient priority, throws forbidden read error", %{
```
eg: "when authenticated with insufficient priority, given admin thread, throws forbidden read error"

unenglishable replied:
please check all other "insufficient priority" tests as well

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+        |> delete(Routes.thread_path(conn, :purge, thread_id), %{})
+        |> json_response(200)
+
+      assert String.slice(response["board_name"], 0..4) == "Board"
```
suggest using `String.starts_with?/2`

unenglishable replied:
https://hexdocs.pm/elixir/1.12/String.html#starts_with?/2

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :global_mod
+    test "after purging thread, does decreases thread posters' post count", %{
```
don't need "does"

please check other uses of "does" in this test file

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    end
+
+    @tag authenticated: :global_mod
+    test "given a valid thread, does unwatch", %{
```
suggestion "does unwatch" -> "unwatches"

this form can be used for other "does" test descriptions

unenglishable reviewed (changes requested):
please review and address comments on `thread_test`

akinsey committed:
test(thread): lock/unlock thread instead of poll

unenglishable commented on test/epochtalk_server_web/controllers/thread_test.exs:
```diff
@@ -181,4 +201,556 @@ defmodule Test.EpochtalkServerWeb.Controllers.Thread do
+    test "given thread that authenticated user moderates, does purge thread", %{
+      conn: conn,
+      thread: %{post: %{thread_id: thread_id}},
+      user: %{id: user_id}
```
https://github.com/epochtalk/epochtalk_server/pull/98#pullrequestreview-2226144176

in response to omitting user from setup return;
you can use this instead:

```users: %{ user: %{id: user_id}}```

akinsey replied:
Is this better than just keeping and returning user for any reason?

akinsey committed:
test(threads): rename threads variable

akinsey committed:
test(thread-controller): fix confusing use of user contexts

akinsey committed:
test(thread-controller): update language to be more correct

unenglishable committed:
test(controllers/thread): refactor test descriptions and variable names

differentiate between thread in admin/superadmin board and thread
created by admin in normal board (used for priority checks)

fix descriptions relating to priority when they were actually related to
permissions

unenglishable reviewed (approved):
personally made a couple more updates

changes approved ✅
